File: src/java/graph/dfs/RecursiveTraversal.java
package graph.dfs;

import java.util.ArrayDeque;
import java.util.Collection;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;


public class RecursiveTraversal {

    public static <T> List<T> traverse(Map<T, List<T>> graph, T root) {
        if (!graph.containsKey(root))
            throw new IllegalArgumentException("Invalid root vertex");

        var explored = new LinkedHashSet<T>();
        
        explore(root, graph, explored);

        // this implementation also explores disconnected subgraphs
        for (var vertex : graph.keySet()) {
            if (!explored.contains(explored)) {
                explore(vertex, graph, explored);
            }
        }

        return List.copyOf(explored);
    }

    private static <T> void explore(T vertex, Map<T, List<T>> graph, Set<T> explored) {
        explored.add(vertex);
        for (var successor : graph.getOrDefault(vertex, List.of())) {
            if (!explored.contains(successor)) {
                explore(successor, graph, explored);
            }
        }
    }


    public static void main(String[] args) {
        var graph = Map.of(
                // first subgraph
                "A", List.of("B", "C", "D"),
                "B", List.of("E"),
                "C", List.of("F", "G"),
                "F", List.of("H", "I"),

                // second subgraph
                "J", List.of("K", "L", "M"),
                "L", List.of("N", "O")
        );

        var result = traverse(graph, "A");

        System.out.println(result); // [A, B, E, C, F, H, I, G, D, .., .., .., .., .., ..]
    }
}

File: src/kotlin/graph/dfs/RecursiveTraversal.kt
package graph.dfs


typealias Graph<T> = Map<T, List<T>>


fun <T> Graph<T>.traverse(root: T): List<T> {
    val graph = this
    require(root in graph) { "Invalid root vertex '$root'" }

    val explored = linkedSetOf<T>()

    fun explore(vertex: T) {
        explored += vertex
        for (successor in graph[vertex].orEmpty()) {
            if (successor !in explored) {
                explore(successor)
            }
        }
    }
    explore(root)

    // this implementation also explores disconnected subgraphs
    for ((vertex, _) in graph) {
        if (vertex !in explored) {
            explore(vertex)
        }
    }
    
    return explored.toList()
}


fun main() {
    val graph = mapOf(
        // first subgraph
        "A" to listOf("B", "C", "D"),
        "B" to listOf("E"),
        "C" to listOf("F", "G"),
        "F" to listOf("H", "I"),

        // second subgraph
        "J" to listOf("K", "L", "M"),
        "L" to listOf("N", "O")
    )

    val result = graph.traverse(root = "A")

    println(result) // [A, B, E, C, F, H, I, G, D, J, K, L, N, O, M]
}